src/caffe/solver.cpp

```diff
@@ -775,9 +776,125 @@ void AdaGradSolver<Dtype>::ComputeUpdateValue(int param_id, Dtype rate) {
   }
 }
 
+
+template <typename Dtype>
+void AdamSolver<Dtype>::ComputeUpdateValue(int param_id, Dtype rate) {
+  const vector<shared_ptr<Blob<Dtype> > >& net_params = this->net_->params();
+  const vector<float>& net_params_lr = this->net_->params_lr();
+  Dtype local_rate = rate * net_params_lr[param_id];
+  const Dtype beta1 = this->param_.beta1();
+  const Dtype beta2 = this->param_.beta2();
+
+  const int t = this->iter_ / this->param_.stepsize() +1;
+  const Dtype step_correction = std::sqrt(Dtype(1)-caffe_ipow(beta2, t))/
+  (Dtype(1.)-caffe_ipow(beta1, t));
+  const int N = net_params[param_id]->count();
+  const Dtype delta = this->param_.delta();
+
+  switch (Caffe::mode()) {
+  case Caffe::CPU: {
+    // update m <- \beta_1 m_{t-1} + (1-\beta_1)g_t
+    caffe_scal(N, Dtype(beta1), this->val_m_[param_id]->mutable_cpu_data());
+    caffe_cpu_axpby(N, Dtype(1-beta1),
+        net_params[param_id]->cpu_diff(), Dtype(0),
+        this->val_t_[param_id]->mutable_cpu_data());
```

Comment:
I don't think this currently does the right thing (val_t_ = (1-beta) net_params + 0*val_t). You can combine the above two lines into one caffe_cpu_axpby using beta1 instead of 0 call.